src/lib.rs

```diff
@@ -1,171 +1,164 @@
-    let fd = unsafe { SOCKET_FD.load(Ordering::Relaxed) };
-    if SOCKET_FD_INIT == fd { return }
-    let mut buf: [u8; 576];
-    buf = unsafe { mem::uninitialized() };
-    let len = {
-        let mut cursor = Cursor::new(&mut buf[..]);
-        let _ = cursor.write_all(unsafe { &PREFIX[..PREFIX_LEN] });
-        for s in strings { let _ = cursor.write_all(s); }
-        cursor.position() as usize
-    };
-    unsafe {
-        libc::send(fd, buf.as_ptr() as *mut c_void, len, MSG_DONTWAIT);
-    }
+pub struct StatsdClient {
+    udp_socket: UdpSocket,
+    prefix: String
 }
 
-use std::cell::RefCell;
+impl StatsdClient {
+    pub fn new(address: &str, prefix: &str) -> Result<StatsdClient, std::io::Error> {
+        let socket = UdpSocket::bind("0.0.0.0:0")?; // NB: CLOEXEC by default
+        socket.set_nonblocking(true)?;
+        socket.connect(address)?;
+        let mut prefix_str = prefix.to_string();
+        prefix_str.push('.');
+        Ok(StatsdClient {
+            udp_socket: socket,
+            prefix: prefix_str
+        })
+    }
 
-#[cfg(target_arch = "x86_64")]
-fn rdtsc() -> u64 {
-    let s: u32;
-    let t: u32;
-    unsafe { asm!("rdtsc" : "={edx}"(s), "={eax}"(t) ::) }
-    ((s as u64) << 32) | (t as u64)
+    pub fn send(&self, strings: &[&str]) {
+        let mut str = String::with_capacity(MAX_UDP_PAYLOAD);
+        str.push_str(&self.prefix);
+        for s in strings { str.push_str(s); }
+        self.udp_socket.send(str.as_bytes());
+    }
 }
 
-fn seed() -> u64 {
-    let seed = 5573589319906701683_u64;
-    let seed = seed.wrapping_mul(6364136223846793005)
-        .wrapping_add(1442695040888963407)
-        .wrapping_add(rdtsc());
-    seed.wrapping_mul(6364136223846793005)
-        .wrapping_add(1442695040888963407)
+#[derive(Debug)]
+pub struct Pcg32Sampling {
+    int_rate: u32,
+    pub buf: String
 }
 
-fn pcg32() -> u32 {
-    thread_local! {
-        static PCG32_STATE: RefCell<u64> = RefCell::new(seed());
+impl Pcg32Sampling {
+    pub fn new(float_rate: f64) -> Box<Pcg32Sampling> {
+        Box::new(Pcg32Sampling {
+            int_rate: (((1.0 - float_rate) * ::std::u32::MAX as f64) as u32),
+            buf: format!("{}", float_rate)})
     }
 
-    PCG32_STATE.with(|state| {
-        let oldstate: u64 = *state.borrow();
-        // XXX could generate the increment from the thread ID
-        *state.borrow_mut() = oldstate.wrapping_mul(6364136223846793005)
-            .wrapping_add(1442695040888963407);
-        ((((oldstate >> 18) ^ oldstate) >> 27) as u32)
-            .rotate_right((oldstate >> 59) as u32)
-    })
-}
-
-pub fn within_rate(r: u32) -> bool {
-    pcg32() > r
+    pub fn accept_next(&self) -> bool {
+        pcg32::random() > self.int_rate
+    }
 }
 
-
-// XXX we need to redo this as a syntax extension that stringifies all
-// the constant arguments so we don't need to format them.  This is
-// particularly important for rate.
 #[macro_export]
 macro_rules! statsd {
```

Comment:
Is there something preventing you to add methods on `StatsdClient` instead of using macros?
It would feel much cleaner to be able to do:
```rust
client.increment("boring", 22, half_rate);
```